File: examples/python/trajectory_generation.py
import sys
import numpy as np

import matplotlib.pyplot as plt

import multicopter_mpc

import pinocchio
import crocoddyl
import example_robot_data

from multicopter_mpc.utils.path import MULTICOPTER_MPC_MULTIROTOR_DIR, MULTICOPTER_MPC_MISSION_DIR, MULTICOPTER_MPC_OCP_DIR

WITHDISPLAY = 'display' in sys.argv
WITHPLOT = 'plot' in sys.argv

uav = example_robot_data.load('iris')
uav_model = uav.model

# # UAV Params
mc_params = multicopter_mpc.MultiCopterBaseParams()
mc_params.fill(MULTICOPTER_MPC_MULTIROTOR_DIR + "/iris.yaml")

# Mission
mission = multicopter_mpc.Mission(uav.nq + uav.nv)
mission.fillWaypoints(MULTICOPTER_MPC_MISSION_DIR + "/hover.yaml")

trajectory = multicopter_mpc.TrajectoryGenerator(uav_model, mc_params, mission)
trajectory.loadParameters(MULTICOPTER_MPC_OCP_DIR + "/trajectory-generator.yaml")

trajectory.createProblem(multicopter_mpc.SolverType.SolverTypeBoxFDDP,
                         multicopter_mpc.IntegratorType.IntegratorTypeEuler, trajectory.dt)
trajectory.setSolverCallbacks(True)

state_guess = mission.interpolateTrajectory("cold")
control = pinocchio.utils.zero(4)
# control += 3.7
control_guess = [control for _ in range(0, len(state_guess) - 1)]
trajectory.setSolverIters(500)
trajectory.setSolverCallbacks(True)
trajectory.solver.th_stop = 1e-5
# trajectory.solve(state_guess, control_guess)
trajectory.solve()


fig, axs = plt.subplots(4, 1)
us = np.vstack(trajectory.controls).T
for idx, ax in enumerate(axs):
    ax.plot(us[idx, :])

plt.show()


state_trajectory = trajectory.states
control_trajectory = trajectory.controls
time_lst = [trajectory.dt for i in range(0, trajectory.n_knots + 1)]

if WITHDISPLAY:
    display = crocoddyl.GepettoDisplay(uav)
    for idx_wp, wp in enumerate(mission.waypoints):
        name = 'world/wp' + str(idx_wp)
        uav.viewer.gui.addXYZaxis(name, [1., 0., 0., 1.], .03, 0.5)
        wp_pose = pinocchio.SE3ToXYZQUATtuple(wp.pose)
        uav.viewer.gui.applyConfiguration(name, wp_pose)

    display.display(state_trajectory, [], [], time_lst, 1)

File: examples/trajectory-generation.cpp
#include "pinocchio/parsers/urdf.hpp"
#include "pinocchio/multibody/model.hpp"

#include "example-robot-data/path.hpp"

#include "crocoddyl/core/solvers/ddp.hpp"

#include "yaml_parser/parser_yaml.h"

#include "multicopter_mpc/multicopter-base-params.hpp"
#include "multicopter_mpc/mission.hpp"
#include "multicopter_mpc/ocp/trajectory-generator.hpp"
#include "multicopter_mpc/path.h"

int main(void) {
  pinocchio::Model model;
  pinocchio::urdf::buildModel(EXAMPLE_ROBOT_DATA_MODEL_DIR "/iris_description/robots/iris_simple.urdf",
                              pinocchio::JointModelFreeFlyer(), model);

  boost::shared_ptr<multicopter_mpc::MultiCopterBaseParams> mc_params =
      boost::make_shared<multicopter_mpc::MultiCopterBaseParams>();
  mc_params->fill(MULTICOPTER_MPC_MULTIROTOR_DIR "/iris.yaml");

  boost::shared_ptr<multicopter_mpc::Mission> mission =
      boost::make_shared<multicopter_mpc::Mission>(model.nq + model.nv);
  mission->fillWaypoints(MULTICOPTER_MPC_MISSION_DIR "/passthrough.yaml");

  boost::shared_ptr<pinocchio::Model> mc_model = boost::make_shared<pinocchio::Model>(model);

  multicopter_mpc::TrajectoryGenerator trajectory(mc_model, mc_params, mission);
  trajectory.loadParameters(MULTICOPTER_MPC_OCP_DIR "/trajectory-generator.yaml");
  trajectory.createProblem(multicopter_mpc::SolverTypes::BoxFDDP, multicopter_mpc::IntegratorTypes::RK4, 1e-2);
  printf("This is the time step %f \n", trajectory.getTimeStep());
  printf("This is the knot number %d \n", trajectory.getKnots());

  trajectory.setSolverCallbacks(true);
  trajectory.solve();
}